### DISCUSSION TITLE
Mock Server Returns HTML Instead of JSON in React Vite

### DISCUSSION THREAD

**Discussion by aramfilak** [mswjs/msw#2235] [Q&A] [answered]
Occasionally, the mock server returns an HTML document instead of the expected JSON object. Initially, the server works as expected, but it randomly returns HTML at unexpected times, disrupting the normal flow of the application.

### Expected Behavior
The mock server should consistently return a JSON object as specified.

### Actual Behavior
The mock server intermittently returns the following HTML document:

```html
<!doctype html>
<html lang="en">
  <head>
    <script type="module">import { injectIntoGlobalHook } from "/@react-refresh";
injectIntoGlobalHook(window);
window.$RefreshReg$ = () => {};
window.$RefreshSig$ = () => (type) => type;</script>

    <script type="module" src="/@vite/client"></script>

    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + React + TS</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/main.tsx?t=1723146340705"></script>
  </body>
</html>

```
### Environment
Mock Server Version: "msw": "^2.3.1",
Client Application: Vite + React + TypeScript + react-query
Browser: chrome

**Comment by aramfilak** [comment]
**Update:** This only happens in the Chrome browser. Since I started using Firefox, the problem has disappeared.

**Reply by itallovidal** [reply]
im having the sabe issue! Both Arc and chrome

**Reply by aramfilak** [reply]
Arc is a Chromium-based browser so i think the is no different between the both.

**Comment by kettanaito** [comment]
Is this timing-sensitive? As in, have you notice this occur after your app has been idle for some time?

**Reply by aramfilak** [reply]
> Is this timing-sensitive? As in, have you notice this occur after your app has been idle for some time?

I think yes, but this also never happens in Firefox, no matter how long the tab hasn't been used and this has nothing to do with React Query or Axios because I tried using the built-in fetch API and also switched to useSWR, but the same problem occurred again

**Reply by kettanaito** [reply, marked as answer]
I suspect this is #2115, #367, #1744. In essence, Chromium has their own Service Worker inactivity policy, and once they terminate MSW's worker, it cannot properly restart unless you refresh the page (reevaluate JavaScript). 

There's one thing we can try on MSW's side but it depends on _how_ Chrome reenables the worker (if it fires any events whatsoever for us to know). I suggest you follow the open issue I referenced for a solution. Thanks.

**Reply by aramfilak** [reply]
@kettanaito  thanks very helpful

**Comment by TinsFox** [comment]
I'm having this problem with Chrome, Arc, and Firefox!

<img width="1508" alt="image" src="https://github.com/user-attachments/assets/19b84146-aac3-4b78-ba60-7d2ae9a8237a">

I don’t know if this issue will be followed up again.

Here is my code https://github.com/TinsFox/mws-example/tree/loader

Hope it helps to solve this problem

**Reply by renejm** [reply]
So it cannot be used with Chrome?

**Reply by TinsFox** [reply]
This should be no problem if you use a server-side framework, I tested with vite. I look forward to your other exploration and sharing in this regard.

**Reply by kettanaito** [reply]
@TinsFox, can you share the steps to reproduce this, please? 

> So it cannot be used with Chrome?

MSW works fine in Chrome. This issue is related to the inactivity policy.

**Reply by TinsFox** [reply]
@kettanaito Have you tried the repository I provided? I haven't been able to control the timing of the recurrence very well, but after logging in and refreshing the interface so that it can request user information, it should be possible to reproduce it.

**Reply by renejm** [reply]
I tried everything possible but couldn't create an endpoint that would return the supposed JSON structure. All I get is an HTML like the one TinsFox reported.
